compiler/damlc/daml-stdlib-src/DA/Exception.daml

```diff
@@ -0,0 +1,78 @@
+-- Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE CPP #-}
+
+#ifndef DAML_EXCEPTIONS
+
+-- | HIDE
+module DA.Exception where
+
+#else
+
+-- | Exception handling.
+module DA.Exception
+    ( AnyException (..)
+    , Exception (..)
+    , displayException
+    , throw
+    , throwPure
+    , ActionCatch (..)
+    ) where
+
+import DA.Record
+import DA.Internal.LF
+import DA.Internal.Exception
+
+instance HasField "message" AnyException Text where
+    getField AnyException{..} = message
+    setField message AnyException{payload} = AnyException{..}
+
+instance HasField "payload" AnyException Any where
+    getField AnyException{..} = payload
+    setField payload AnyException{message} = AnyException{..}
+
+-- | Exception typeclass. This should not be implemented directly,
+-- instead, use the `exception` syntax.
+class Exception t where
+    toAnyException : t -> AnyException
+    fromAnyException : AnyException -> Optional t
+
+instance Exception AnyException where
+    toAnyException = identity
+    fromAnyException = Some
+
+-- | Get the error message associated with an exception.
+displayException : Exception e => e -> Text
+displayException e = getField @"message" (toAnyException e)
+
+-- | Throw an exception.
+throw : (Action m, Exception e) => e -> m t
+throw e = do
+    pure ()
+    throwPure e
```

Comment:
This function worries me a bit. You are still tying it to evaluation not execution, you just rely on the fact that `>>=` won’t evaluate the second part but since the first one here is known, it absolutely could and a naive free monad will do exactly that. In DAML Script we hacked around this by throwing the free monad in another `State Token` and we pass in the token from the free monad interpreter which mostly solves this for `trace` statements but I’d prefer not to rely on this too much.

Have you considered defining a `MonadThrow` typeclass instead so that for example DAML Script could map this to a `throw` primitive which we interpret in the free monad interpreter instead of relying on evaluation happening at the right time?

Reply:
Ah, that's a good idea. :)

Reply:
Ok, I added `ActionThrow`, and changed `ActionCatch` to use `ActionThrow` as superclass (since we can implement `ActionThrow` anywhere, and it would be annoying to require both constraints if you wanted to both throw and catch exceptions generically). :-)

Reply:
Thanks!